swap/swap.go

```diff
@@ -314,22 +314,52 @@ func createOwner(prvkey *ecdsa.PrivateKey) *Owner {
 	}
 }
 
-// Add is the (sole) accounting function
-// Swap implements the protocols.Balance interface
-func (s *Swap) Add(amount int64, peer *protocols.Peer) (err error) {
+// checkBalance checks that the amount would not result in crossing the disconnection threshold
+func (s *Swap) checkBalance(amount int64, peer *protocols.Peer) (err error) {
 	swapPeer := s.getPeer(peer.ID())
 	if swapPeer == nil {
 		return fmt.Errorf("peer %s not a swap enabled peer", peer.ID().String())
 	}
-	swapPeer.lock.Lock()
-	defer swapPeer.lock.Unlock()
 
 	// check if balance with peer is over the disconnect threshold and if the message would increase the existing debt
 	balance := swapPeer.getBalance()
 	if balance >= s.params.DisconnectThreshold && amount > 0 {
 		return fmt.Errorf("balance for peer %s is over the disconnect threshold %d and cannot incur more debt, disconnecting", peer.ID().String(), s.params.DisconnectThreshold)
 	}
 
+	return nil
+}
+
+// Check is called as a *dry run* before applying the actual accounting to an operation.
+// It only checks that performing a given accounting operation would not incur in an error.
+// If it returns no error, this signals to the caller that the operation is safe
+func (s *Swap) Check(amount int64, peer *protocols.Peer) (err error) {
+	swapPeer := s.getPeer(peer.ID())
+	if swapPeer == nil {
+		return fmt.Errorf("peer %s not a swap enabled peer", peer.ID().String())
+	}
+
+	swapPeer.lock.Lock()
+	defer swapPeer.lock.Unlock()
+	// currently this is the only real check needed:
+	return s.checkBalance(amount, peer)
+}
+
+// Add is the (sole) accounting function
+// Swap implements the protocols.Balance interface
+func (s *Swap) Add(amount int64, peer *protocols.Peer) (err error) {
+	swapPeer := s.getPeer(peer.ID())
+	if swapPeer == nil {
+		return fmt.Errorf("peer %s not a swap enabled peer", peer.ID().String())
+	}
+
+	swapPeer.lock.Lock()
+	defer swapPeer.lock.Unlock()
+	// we should probably check here again:
```

Comment:
Why? With the peer under lock the balance couldn't have changed since the call to `Check`, right?

Reply:
That is correct. However, just the nature of the two being separate functions *could* lead to the situation that `Add` be called in a different workflow. For example, I do not recall if all tests now do this `Validate-Apply` workflow - some tests might be calling directly `Add` (I would actually assume so). 

Even if the probability is very low, it is correct to check again. However, I am fine removing the check if a majority thinks it is superfluous.

Reply:
Hmm. I don't really like this double-check approach. From just looking at it, it looks as if you are doing something redundantly and it is probably also not water-tight.

A different solution could be that the Check function would leave a trace in Swap that a certain amount has been checked and Add only executes on a validated amount. For example, we could create a mapping where each amount maps to a bool (Checked)), to be set to true during the function `Check` and toggled again after `Add` has executed.

Might be too much engineering though. Let me know your thoughts!

Reply:
Removing the additional check, and adding a comment that it is the responsibility of the calling function to check the balance would be more than enough, I think though!